### DISCUSSION TITLE
Shorthand for `Enumerable.Empty<T>()` and `Array.Empty<T>()`

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6779] [Language Ideas] [answered]
As it stands, there is no quick and convenient way to declare that you want an empty non-null enumerable collection be returned in a method. And the most common approach people may take is `new T[0]` which will trigger a suggestion and provide a code fix to change it into `Array.Empty<T>()`, however it sitll feels clunky and unintuitive to be required to type out the entire type of the empty collection, when you're directly returning it.

Consider the following (trivial) example:
```csharp
public IEnumerable<T> FilterOrEmpty<T>(IEnumerable<T>? other, Func<T, bool> predicate)
{
    return other?.Where(predicate) ?? Enumerable.Empty<T>();
}
```

There are two qualifiers for specifying that there must be an empty enumerable, and the qualifier of the type itself specified as a generic argument, when the information about an empty enumerable could be trivially expressed as `empty`, like so:

```csharp
public IEnumerable<T> FilterOrEmpty<T>(IEnumerable<T>? other, Func<T, bool> predicate)
{
    return other?.Where(predicate) ?? empty;
}
```

In this case due to type inference, `empty` target-types to `T`, when in its full form it would be written out as `empty(T)` like `default(T)`.

Additionally, for arrays the proposed variant is `empty[]` for target-typed empty arrays, or `empty T[]` substituting `Array.Empty<T>()`, which acts similarly to `new T[0]`.

**Comment by bernd5** [comment, marked as answer]
See https://github.com/dotnet/csharplang/issues/5354

**Reply by jnm2** [reply]
Where the common way to create any kind of empty collection with target typing will probably be `[]`, parallel with today's `is []`.
